python_modules/dagster/dagster/_core/definitions/asset_daemon_cursor.py

```diff
@@ -15,26 +17,116 @@
     TimeWindowPartitionsDefinition,
     TimeWindowPartitionsSubset,
 )
-from dagster._serdes.serdes import deserialize_value, serialize_value
+from dagster._serdes.serdes import PackableValue, deserialize_value, serialize_value
 
 from .asset_graph import AssetGraph
 from .asset_subset import AssetSubset
-from .partition import PartitionsSubset
+from .partition import PartitionsDefinition, PartitionsSubset
 
 if TYPE_CHECKING:
-    from .asset_condition import AssetConditionEvaluation
+    from .asset_condition import AssetCondition, AssetConditionEvaluation, AssetConditionSnapshot
+    from .asset_condition_evaluation_context import AssetConditionEvaluationContext
 
+ExtrasDict = Mapping[str, PackableValue]
 
-class AssetDaemonAssetCursor(NamedTuple):
+T = TypeVar("T")
+
+
+def _get_placeholder_missing_condition() -> "AssetCondition":
+    """Temporary hard-coding of the hash of the "materialize on missing" condition. This will
+    no longer be necessary once we start serializing the AssetDaemonCursor.
+    """
+    from .asset_condition import RuleCondition
+    from .auto_materialize_rule import MaterializeOnMissingRule
+
+    return RuleCondition(MaterializeOnMissingRule())
+
+
+_PLACEHOLDER_HANDLED_SUBSET_KEY = "handled_subset"
+
+
+class AssetConditionCursorExtras(NamedTuple):
+    """Class to represent additional unstructured information that may be tracked by a particular
+    asset condition.
+    """
+
+    condition_snapshot: "AssetConditionSnapshot"
```

Comment:
Why put this on this object instead of on the asset condition cursor?

Reply:
The idea is that the AssetConditionCursor is _not_ a recursive object -- it just has a flat set of properties.

One of those properties is a list of these extras objects, which are mapped to a specific sub-expression. So the condition_snapshot here is the snapshot of one of the child conditions, whereas the condition_snapshot on the cursor is the snapshot of the root condition.

Another option would be to just formalize this mapping structure and do one of those NonScalarSerializableMapping things where we have a mapping from ConditionSnapshot to Mapping[str, PackableValue].

That might be a cleaner solution as it avoids adding an extra whitelisted class

Reply:
Ahh I see. I think I got tripped up by different interpretations of the plural "extras".

I do think a mapping would make it clearer. Putting `descendants` in the name of the attribute on `AssetConditionCursor` would also make it clearer.